core/src/main/scala/cats/derived/hash.scala

```diff
@@ -1,91 +1,91 @@
+      case Nil => MurmurHash3.finalizeHash(hash, length)
     }
 
-    def eqv(x: H :: T, y: H :: T): Boolean =
-      hashH.unify.eqv(x.head, y.head) && hashT.eqv(x.tail, y.tail)
+    loop(hashes(x), seed, 0)
   }
 }
 
-trait MkHash0 extends MkHash1 {
-  implicit def deriveHashCaseObject[A](
-                                      implicit repr: Generic.Aux[A, HNil]): MkHash[A] = new MkHash[A] {
-    def hash(x: A): Int = x.hashCode
+private[derived] object HashBuilder {
+  import shapeless._
 
-    def eqv(x: A, y: A): Boolean = true
-  }
+  implicit val hashBuilderHNil: HashBuilder[HNil] =
+    instance(_ => Nil, (_, _) => true)
 
-}
+  implicit def hashBuilderHCons[H, T <: HList](
+    implicit H: Hash[H] OrElse MkHash[H], T: HashBuilder[T]
+  ): HashBuilder[H :: T] = instance(
+    { case h :: t => H.unify.hash(h) :: T.hashes(t) },
+    { case (hx :: tx, hy :: ty) => H.unify.eqv(hx, hy) && T.eqv(tx, ty) }
+  )
 
-trait MkHash1 {
-  implicit def fromBuilder[A](implicit builder: HashBuilder[A]): MkHash[A] = new MkHash[A] {
-    override def hash(x: A): Int = {
-      val hashes = builder.hashes(x)
-      runtime.Statics.finalizeHash(hashes.foldLeft(-889275714)(runtime.Statics.mix), hashes.length)
+  private def instance[A](f: A => List[Int], g: (A, A) => Boolean): HashBuilder[A] =
+    new HashBuilder[A] {
+      def hashes(x: A) = f(x)
+      def eqv(x: A, y: A) = g(x, y)
     }
+}
 
-    override def eqv(x: A, y: A): Boolean = builder.eqv(x, y)
-  }
+private[derived] abstract class MkHashDerivation extends MkHashGeneric {
+  import shapeless._
+
+  implicit val mkHashCNil: MkHash[CNil] =
+    instance(_ => 0, (_, _) => true)
+
+  implicit def mkHashCCons[L, R <: Coproduct](
+    implicit L: Hash[L] OrElse MkHash[L], R: MkHash[R]
+  ): MkHash[L :+: R] = instance({
+    case Inl(l) => L.unify.hash(l)
+    case Inr(r) => R.hash(r)
+  }, {
+    case (Inl(lx), Inl(ly)) => L.unify.eqv(lx, ly)
+    case (Inr(rx), Inr(ry)) => R.eqv(rx, ry)
+    case _ => false
+  })
+
+  implicit def mkHashCaseObject[A](implicit A: Generic.Aux[A, HNil]): MkHash[A] =
+    instance(_.hashCode, (_, _) => true)
+}
 
-  implicit def emptyCoproductDerivedHash: MkHash[CNil] = null
-  // used when Hash[V] (a member of the coproduct) has to be derived.
-  implicit def coproductDerivedHash[L, R <: Coproduct](
-     implicit
-     hashV: Hash[L] OrElse MkHash[L],
-     hashT: MkHash[R]): MkHash[L :+: R] = new MkHash[L :+: R] {
-    def hash(value: L :+: R): Int = value match {
-      case Inl(l) => hashV.unify.hash(l)
-      case Inr(r) => hashT.hash(r)
+private[derived] abstract class MkHashGeneric {
+  import shapeless._
+
+  implicit def mkHashGenericProduct[A, R <: HList](
+    implicit A: Generic.Aux[A, R], R: Lazy[HashBuilder[R]], ev: A <:< Product = null
+  ): MkHash[A] = instance(
+    x => R.value.hash(A.to(x), if (ev == null) MurmurHash3.productSeed else util.VersionSpecific.productSeed(x)),
```

Comment:
I'm not sure about this to be honest.

Reply:
Could we use a pattern match here? like 
```scala
x match {
   case p : Product => util.VersionSpecific.productSeed(p)
   case _ => MurmurHash3.productSeed
}
```

Reply:
Yes, that's the part I'm not sure about. On one hand, `ev: A <:< Product = null` looks like a hack, but on the other pattern matching on the type parameter would break parametricity.

But what if the solution is to drop this law altogether from `cats-testkit`?

Reply:
I need to think about the law. Could we split this one method into two derivation methods with different priorities? One for Product another for the rest.

Reply:
Sure, that's also an option :thumbsup:

Reply:
Done. We can later remove the version specific code if we decide not to enforce this law.